Include discussion of embedded skips (#1764)
r-lib/testthat: vignettes/skipping.Rmd

As of testthat 3.0.0, ProgressReporter (used interactively) and CheckReporter (used inside of `R CMD check`) also display a summary of skips across all tests.
It looks something like this:

```         
── Skipped tests  ───────────────────────────────────────────────────────
● No token (3)
● On CRAN (1)
```

You should keep an on eye this when developing interactively to make sure that you're not accidentally skipping the wrong things.

## Helpers

If you find yourself using the same `skip_if()`/`skip_if_not()` expression across multiple tests, it's a good idea to create a helper function.

> This function should start with `skip_` and live in a `test/helper-{something}.R`:

Reviewer: ```suggestion
This function should start with `skip_` and live in a `test/helper-{something}.R` file:
```

```{r}

> skip_if_dangerous <- function() {
>   if (!identical(Sys.getenv("DANGER"), "")) {
>     skip("Not run in dangerous environments.")
>   } else {
>     invisible()

Reviewer: Connecting back to my nitpick comment on the other PR, I find negated conditions sort of mind-bendy. Can we do it the other (more natural) way around?

```suggestion
skip_if_dangerous <- function() {
  if (identical(Sys.getenv("DANGER"), "")) {
    invisible()
  } else {
    skip("Not run in dangerous environments.")
```

Another nice example:

```
skip_if_not_utf8 <- function() {
  skip_if_not(l10n_info()$`UTF-8`)
}
```

Author: Let's discuss that more there, because I find the current structure more clear.

  } else {
    if (is_testing()) {
      testthat::skip("Pandoc not available")
    } else {
      abort("Pandoc not available")
    }
  }
  
  ...
}
```

If Pandoc is not available when `convert_markdown_to_html()` executes, it throws an error *unless* it appears to be part of a test run, in which case the test is skipped.
This is an alternative to implementing a custom skipper, e.g. `skip_if_no_pandoc()`, and inserting it into many of pkgdown's tests.

We don't want pkgdown to have a runtime dependency on testthat, so pkgdown includes a copy of `testthat::is_testing()`:

```{r eval = FALSE}
is_testing <- function() {
  identical(Sys.getenv("TESTTHAT"), "true")
}
```

It might look like the code appears to still have a runtime dependency on testthat, because of the call to `testthat::skip()`.
But `testthat::skip` is only ever run when you're inside a test, where you can assume that testthat is installed.

We have mixed feelings about this approach.
On the one hand, it feels elegant and concise, and it absolutely guarantees that you'll never miss a needed skip in one of your tests.

> On the other hand, it mixed code and tests in an unusual way, and it when you're focused on the tests, the existence of a skip in code is easy to miss.

Reviewer: ```suggestion
On the other hand, it mixes code and tests in an unusual way, and when you're focused on the tests, it's easy miss the fact that a package function contains a `skip()`.
```
